platforms/core-runtime/internal-instrumentation-processor/src/main/java/org/gradle/internal/instrumentation/extensions/property/PropertyUpgradeAnnotatedMethodReader.java

```diff
@@ -364,34 +368,52 @@ private AccessorSpec getAccessorSpec(
         );
     }
 
-    private static Type extractOriginalType(ExecutableElement method, AnnotationMirror annotation) {
+    private static TypeName extractOriginalType(ExecutableElement method, AnnotationMirror annotation) {
         Optional<? extends AnnotationValue> annotationValue = AnnotationUtils.findAnnotationValue(annotation, "originalType");
-        Type type = annotationValue.map(v -> extractType((TypeMirror) v.getValue())).orElse(DEFAULT_TYPE);
-        if (!type.equals(DEFAULT_TYPE)) {
-            return type;
+        TypeName typeName = annotationValue.map(v -> v.getValue() instanceof DeclaredType
+            // We use DeclaredType.asElement().asType() so if the original type is a parametrized type,
+            // e.g. Iterable<T>, resolved TypeName contains information that it's Iterable<T> and not just Iterable
+            ? TypeName.get(((DeclaredType) annotationValue.get().getValue()).asElement().asType())
+            : TypeName.get((TypeMirror) annotationValue.get().getValue())
+        ).orElse(DEFAULT_TYPE);
+        if (!typeName.equals(DEFAULT_TYPE)) {
+            return typeName;
         }
         return extractOriginalTypeFromGeneric(method, method.getReturnType());
     }
 
-    private static Type extractOriginalTypeFromGeneric(ExecutableElement method, TypeMirror typeMirror) {
+    private static TypeName extractOriginalTypeFromGeneric(ExecutableElement method, TypeMirror typeMirror) {
         String typeName = method.getReturnType() instanceof DeclaredType
             ? ((DeclaredType) method.getReturnType()).asElement().toString()
             : method.getReturnType().toString();
         GradleLazyType gradleLazyType = GradleLazyType.from(typeName);
         switch (gradleLazyType) {
             case CONFIGURABLE_FILE_COLLECTION:
-                return FILE_COLLECTION.asType();
+                return FILE_COLLECTION.asClassName();
             case DIRECTORY_PROPERTY:
             case REGULAR_FILE_PROPERTY:
-                return Type.getType(File.class);
+                return ClassName.get(File.class);
             case LIST_PROPERTY:
-                return Type.getType(List.class);
+                return ParameterizedTypeName.get(ClassName.get(List.class),
+                    TypeName.get(((DeclaredType) typeMirror).getTypeArguments().get(0))
```

Comment:
💅 Let's extract this to some utility method.

Reply:
Done